### DISCUSSION TITLE
SEO problem, canonical URL's are always redirected

### DISCUSSION THREAD

**Discussion by mikevercoelen** [amannn/next-intl#1035] [Q&A]
Hi there, awesome library, we're building a reasonably sized e-commerce platform in the EU using this, and we're very happy with it.

We're currently stuck on an SEO problem and I hope someone can help us out.

We use `localePrefix: 'always'` + localized path names + `defaultLocale: 'cs' (Czech)` (not English, should we?)

This is the issue:

- Since locale detection is based on a cookie, a crawler will ALWAYS get a redirect, which is bad for SEO, how do we fix this?

i.e. `curl -L -H "Accept-Language: cs-CZ" -H "Cookie: NEXT_LOCALE=cs" https://example.com/cs` this works correctly, no redirect, however a crawler does not have a cookie.

This also messes up canonical URL's, since a crawler follows a canonical URL, but since there's no cookie, it's redirected and canonical URL's should never have a redirect, which is considered bad.

Anyone got a clue?

Peace.

**Comment by mikevercoelen** [comment]
To follow up on this:

1. Why is it redirecting if the NEXT_LOCALE cookie and accept-header is equal to the default locale? It should NOT redirect in that case? Or alternatively, it would be good if this would be configurable?
2. It would be great if we can configure disabling locale detection when the request comes from a crawler

**Comment by amannn** [comment]
Hey @mikevercoelen, glad to hear if next-intl works well for you so far! 🙌

> Since locale detection is based on a cookie, a crawler will ALWAYS get a redirect, which is bad for SEO, how do we fix this?

I guess you mean at the root `/`, right? If a crawler visits a prefixed route like `/cs`, then the locale will be derived from the pathname, regardless of the `accept-language` header and the presence/absence of a cookie.

See e.g.:

```
# Visit pathname with non-default locale as prefix
curl -v -s -D - -o /dev/null https://next-intl-example-app-router.vercel.app/de 
# Result: Status code 200, set-cookie: NEXT_LOCALE=de
```

> This also messes up canonical URL's, since a crawler follows a canonical URL, but since there's no cookie, it's redirected and canonical URL's should never have a redirect, which is considered bad.

You might be interested in this discussion regarding canonical URLs: https://github.com/amannn/next-intl/discussions/629. The tldr; is that you shouldn't use canonical links but alternate links, which is what `next-intl` does out-of-the-box.

> Why is it redirecting if the NEXT_LOCALE cookie and accept-header is equal to the default locale? It should NOT redirect in that case? Or alternatively, it would be good if this would be configurable?

Sorry, I can't follow. Can you provide an example curl for https://next-intl-example-app-router.vercel.app that shows this behavior?

> It would be great if we can configure disabling locale detection when the request comes from a crawler

You can generally [compose the middleware](https://next-intl-docs.vercel.app/docs/routing/middleware#composing-other-middlewares) and add any kind of custom functionality on top if you like.

For convenience, `next-intl` provides a redirect at the root `/` for the best matching locale, which is considered good practice. However, you can also provide a custom page for this route, e.g. to show a language prompt to the user. In that case, you can remove `/` from the middleware matcher and add a `app/page.tsx`. Note however, that you can not use `next-intl` on this route then since no locale is known and you probably have to add basic page markup like `<html>` if you define this in `app/[locale]/layout.tsx` for localized pages.

Hope this helps!
